sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -392,4 +392,267 @@ case class GpuArrayMax(child: Expression) extends GpuBaseArrayAgg with ImplicitC
+            withResource(difference.greaterOrEqualTo(scalarZero)) { pred =>
+              pred.ifElse(one, negativeOne)
+            }
+          }
+        }
+        val size = closeOnExcept(step) { _ =>
+          withResource(difference.abs()) { absDifference =>
+            absDifference.add(one)
+          }
+        }
+        Seq(size, step)
+      }
+    }
+  }
+
+  override def doColumnar(start: GpuColumnVector, stop: GpuColumnVector): ColumnVector = {
+    withResource(calculateSizeAndStep(start.getBase, stop.getBase, start.dataType())) { ret =>
+      ColumnVector.sequence(start.getBase, ret(0), ret(1))
+    }
+  }
+
+  override def doColumnar(start: GpuScalar, stop: GpuColumnVector): ColumnVector = {
+    withResource(calculateSizeAndStep(start.getBase, stop.getBase, stop.dataType())) { ret =>
+      withResource(ColumnVector.fromScalar(start.getBase, stop.getRowCount.toInt)) { startV =>
+        ColumnVector.sequence(startV, ret(0), ret(1))
+      }
+    }
+  }
+
+  override def doColumnar(start: GpuColumnVector, stop: GpuScalar): ColumnVector = {
+    withResource(calculateSizeAndStep(start.getBase, stop.getBase, start.dataType())) { ret =>
+      ColumnVector.sequence(start.getBase, ret(0), ret(1))
+    }
+  }
+
+  override def doColumnar(numRows: Int, lhs: GpuScalar, rhs: GpuScalar): ColumnVector =
+    throw new IllegalStateException("This is not supported yet")
+
+}
+
+/** GpuSequence with step */
+case class GpuSequenceWithStep(start: Expression, stop: Expression, step: Expression,
+    timeZoneId: Option[String] = None) extends GpuTernaryExpression with TimeZoneAwareExpression
+    with GpuSequenceTrait {
+
+  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression =
+    copy(timeZoneId = Some(timeZoneId))
+
+  override def first: Expression = start
+
+  override def second: Expression = stop
+
+  override def third: Expression = step
+
+  override def dataType: DataType = ArrayType(start.dataType, containsNull = false)
+
+  /* TODO do we need to check below conditions?
+    (step > num.zero && start <= stop)
+        || (step < num.zero && start >= stop)
+        || (step == num.zero && start == stop),
+   */
+  private def calculateSize(
+      start: BinaryOperable,
+      stop: BinaryOperable,
+      step: BinaryOperable,
+      rows: Int,
+      dt: DataType): ColumnVector = {
+    // First, calculate sizeWithNegative=floor((stop-start)/step)+1.
+    //     if step = 0, the div operation in cudf will get MIN_VALUE, which is ok for this case,
+    //     since when size < 0, cudf will not generate sequence
+    // Second, calculate size = if(sizeWithNegative < 0) 0 else sizeWithNegative
+    // Third, if (start == stop && step == 0), let size = 1.
+    withResource(stop.sub(start)) { difference =>
+      withResource(difference.floorDiv(step)) { quotient =>
+        withResource(numberScalar(dt, 1)) { one =>
+          withResource(quotient.add(one)) { sizeWithNegative =>
```

Comment:
When `withResource` nests this deeply, that's usually an indication that we're holding onto one or more GPU results longer than necessary, adding undesired and avoidable memory pressure.  For example, we compute `quotient` here and only need it to compute `sizeWithNegative`, yet we hold onto the GPU memory for the `quotient` result until after the entire calculation completes.  The memory can be freed earlier with something like this:
```scala
  withResource(numberScalar(dt, 1)) { one =>
    val sizeWithNegative = withResource(difference.floorDiv(step)) { quotient =>
      quotient.add(one)
    }
    withResource(sizeWithNegative) { sizeWithNegative =>
      ....
```

Reply:
Thx @jlowe. Changed this accordingly.